File: Data Structures/comp-geometry/convexhull.cpp
#include <iostream>
#include <vector>

using namespace std;
/*
A convex hull is the smallest convex polygon that contains a given set of points. It is a useful concept in computational geometry and has applications in various fields such as computer graphics, image processing, and collision detection.

A convex polygon is a polygon in which all interior angles are less than 180 degrees. A convex hull can be constructed for any set of points, regardless of their arrangement.*/

 


// Orientation function
/*
comparing two slopes we get the orientation of the p3 wrt to line p1p2

d(p32)>d(p12) -> cc
d(p32)<d(p12) -> clock
d(32)=d(12) -> collinear



*/

// Polar angle, arctan2

void graham_scan(vector<int,int> points[]){
    //choosing minimum point for initiating
    vector<int,int> p0 = min(points);
    // sorting based on points
    points.sort();

}

vector<int,int> min(vector<int,int> points[]){
    vector<int,int> min_point = points[0];

    for(int i = 0; i < points->size(); i++){
        if (points[i] > min_point) min_point = points[i];
    }

    return min_point;
}

void main(){


// Sort points

// Accept or Reject Points


}

File: Data Structures/comp-geometry/convexhull.py
from math import atan2, sqrt, pi

def polar_angle(p1,p2):
    if p1[1] == p2[1]:
        return -pi
    dy = p1[1] - p2[1]
    dx = p1[0] - p2[0]
    return atan2(dy,dx)

def dist(p1,p2):
    x1,y1,x2,y2 = *p1, *p2
    return (sqrt((y2-y1)**2 + (x2-x1)**2))

def orientation(p0,p1,p2):
    x1,y1,x2,y2,x3,y3 = *p0,*p1,*p2
    slope_comp = (y3-y2)*(x2-x1) - (y2-y1)*(x3-x2)
    
    if slope_comp > 0: return 1
    elif slope_comp < 0: return -1
    elif slope_comp == 0: return 0

def graham_scan(points):
    p0 = min(points, key=lambda p: (p[1], p[0]))
    points.sort(key=lambda p: (polar_angle(p0,p), dist(p0,p)))
    hull = []
    for i in range(len(points)):
        while len(hull) >= 2 and orientation(hull[-2], hull[-1], points[i]) != 1:
            print(f"Checking: {hull}")
            hull.pop()
        print(f"Checked:{hull}")
        hull.append(points[i])
    return hull



points = [[1,2],[2,2],[4,2]]
hull = graham_scan(points=points)
print(hull)